verilog/tools/ls/autoexpand.cc

```diff
@@ -243,27 +362,265 @@ void AutoExpander::Module::EmitPortConnections(
     const absl::string_view header,
     const std::function<bool(const Port &)> &pred, const Template *tmpl) const {
   bool first = true;
+  GenerateConnections(
+      tmpl, [&](const Port &port, const absl::string_view connected_name) {
+        if (!pred(port)) return;
+        if (first) {
+          if (output.tellp() != 0) output << ',';
+          output << '\n' << indent << "// " << header;
+          first = false;
+        } else {
+          output << ',';
+        }
+        output << '\n'
+               << indent << '.' << port.name << '(' << connected_name << ")";
+      });
+}
+
+void AutoExpander::Module::EmitUndeclaredInputDeclarations(
+    std::ostream &output, const absl::string_view indent) const {
+  EmitPortBodyDeclarations(output, indent, [](const Port &port) {
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Input;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredInoutDeclarations(
+    std::ostream &output, const absl::string_view indent) const {
+  EmitPortBodyDeclarations(output, indent, [](const Port &port) {
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Inout;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredOutputDeclarations(
+    std::ostream &output, const absl::string_view indent) const {
+  EmitPortBodyDeclarations(output, indent, [](const Port &port) {
+    return port.declaration == Port::Undeclared &&
+           port.direction == Port::Output;
+  });
+}
+
+void AutoExpander::Module::EmitUndeclaredOutputWireDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view auto_span) const {
+  absl::flat_hash_set<absl::string_view> declared_wires;
+  for (const auto &reg : FindAllNetVariables(symbol)) {
+    const SyntaxTreeLeaf *const net_name_leaf =
+        GetNameLeafOfNetVariable(*reg.match);
+    const absl::string_view net_name = net_name_leaf->get().text();
+    if (net_name.end() <= auto_span.begin() &&
```

Comment:
Maybe a little helper function `static bool OutsideSpan(absl::string_view piece, absl::string_view span);` for this test might be useful. Improves readability, and you need it around 448 again.